header: return err from reader.Peek()

igorwwwwwwwwwwwwwwwwwwww opened the pull request:
The current behaviour of `Read` is to discard the underlying error and return `ErrNoProxyProtocol`. That makes it tricky to debug what is wrong.

In my case, I was trying to nest `tls.NewListener` inside a `proxyproto.Listener` but got the ordering wrong, the underlying error was:

```
tls: first record does not look like a TLS handshake
```

This patch surfaces that error to the caller.

Note: It may be a good idea to do the same thing with the other `.Peek()` calls too.

pires commented:
Hey @igorwwwwwwwwwwwwwwwwwwww! Thanks for the report and patch. Could you add a test that mimics your scenario so we see the TLS error actually bubbling up? And since we're at it, could you also please change all other `Peek()` calls to return an error?

igorwwwwwwwwwwwwwwwwwwww commented:
@pires I've added a test with a reader that returns a custom error.

I also added some logic to handle EOF differently. Some tests are expecting EOF to return `ErrNoProxyProtocol`, and that seems reasonable.

The error handling got a bit gnarly, so if you have suggestions for how to refactor this, I'm happy to do so.

pires commented on header.go:
```diff
@@ -123,10 +123,34 @@ func (header *Header) TLVs() ([]TLV, error) {
+		if err == io.EOF {
+			return nil, ErrNoProxyProtocol
+		}
+		return nil, err
```
IIUC this is the fix to the problem you were observing, right?

igorwwwwwwwwwwwwwwwwwwww replied:
That is correct.

pires commented on header.go:
```diff
@@ -123,10 +123,34 @@ func (header *Header) TLVs() ([]TLV, error) {
+			if err == io.EOF {
+				return nil, ErrNoProxyProtocol
+			}
+			return nil, err
```
What type of errors do you think may happen here that can't be wrapped w/ a `ErrNoProxyProtocol`?

igorwwwwwwwwwwwwwwwwwwww replied:
The underlying `Read` syscall could fail afaik. In practice this is unlikely to happen here because we're using bufio with a default buffer size of 4k. So I'd expect the first `Read` to have filled the buffer with at least 6 bytes. So this is mostly handling a theoretical case.

pires replied:
Gotcha.

pires commented on header_test.go:
```diff
@@ -41,6 +50,18 @@ func TestReadTimeoutV1Invalid(t *testing.T) {
+	} else if err != errReadIntentionallyBroken {
+		t.Fatalf("expected error %s, actual %s", errReadIntentionallyBroken, err)
+	}
+}
```
This is OK but I'd be happier w/ a test case that covers the real scenario you ran into, the TLS issue.

pires commented:
👏  https://github.com/pires/go-proxyproto/releases/tag/0.1.2
